server/aesdsocket.c

```diff
@@ -0,0 +1,377 @@
+		pid = fork();
+		if (-1 == pid)
+		{
+			perror("fork()");
+			exit(ERROR_CODE);
+		}
+		else if (pid != 0 )
+		{
+			
+			exit(EXIT_SUCCESS);
+		}
+
+        /* The child process continues execution and performs the following steps to daemonize itself:*/
+		/* create new session and process group */
+        /* Detaching the process from the controlling terminal and creating a new session leader*/
+		if(-1 == setsid()) 
+		{
+			perror("setsid()");
+			exit(ERROR_CODE);
+        }
+
+		/* set the working directory to the root directory */
+        /* This is done to ensure that the daemon process does not have any specific working directory tied to the terminal */
+		if (-1 == chdir("/"))
+		{
+			exit(ERROR_CODE);
+		}
+
+       
+        /* It redirects the standard input, output, and error to /dev/null by opening and duplicating file descriptors, 
+         * effectively detaching the process from the terminal and eliminating any input or output connections with the daemon.
+         */
+        /* redirect fd's 0,1,2 to /dev/null */
+		open ("/dev/null", O_RDWR);     /* stdin */
+        dup (0);                        /* stdout */
+        dup (0);                        /* stderror */
+	}    
+
+    //Create file to push data to /var/tmp/aesdsocketdata
+    data_file_fd = open(DATA_FILE,O_CREAT|O_RDWR,0777);
+	if(-1 == data_file_fd)
+	{
+	    perror("Error: Couldn't open the file at /var/temp/aesdsocketdata");
+		exit(ERROR_CODE);
+	}
+
+    if(-1 == listen(main_sockfd,BACKLOG)) //backlog assumed 10 i.e, >1
+    {
+        perror("listen()");
+        freeaddrinfo(res);
+        close(main_sockfd);
+        exit(ERROR_CODE);
+    }
+    freeaddrinfo(res); // all done with this structure
+    /*
+     *    Need a while(1) so that we can accept many client requests and handle them. 
+     */
+    while(1)
+    {
+        socklen_t client_len = sizeof(client_addr);
+        client_sockfd = accept(main_sockfd,(struct sockaddr*)&client_addr,&client_len);
+
+        if (-1 == client_sockfd) 
+        {
+            perror("accept()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+
+        /*
+         *  Reference for  inet_ntop : https://man7.org/linux/man-pages/man3/inet_ntop.3.html
+         */ 
+        if(NULL == inet_ntop(AF_INET, get_in_addr((struct sockaddr*)&client_addr),ip_addr, sizeof(ip_addr)))
+        {
+            perror("inet_ntop()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+        else
+        {
+            syslog(LOG_DEBUG,"Accepted connection from %s", ip_addr);
+        }
+
+        char *data_buf = (char*)calloc(initial_alloc_size,sizeof(char));
+        if(NULL == data_buf)
+        {
+            perror("calloc()");
+            error_handler();
+            exit(ERROR_CODE);
+        }
+
+        /*
+         * For completing any open connection operations i.e., resv() and send(), need to block the signal using sigprocmask()
```

Comment:
@Rishikesh38 this design also means that you can't use the signal to unblock the socket operation, however, and might be related to the issue with Ctrl->C you mentioned in the signal handler.

Reply:
ok, Will use the shutdown function instead of block and unblock